tests/ModelContextProtocol.Tests/Server/McpServerResourceTemplateTests.cs

```diff
@@ -0,0 +1,544 @@
+            Assert.Same(mockServer.Object, server);
+            return new List<string>() { "42", "43" };
+        }, new() { Name = "Test" });
+        var result = await resource.ReadAsync(
+            new RequestContext<ReadResourceRequestParams>(mockServer.Object) { Params = new() { Uri = "resource://Test" } },
+            TestContext.Current.CancellationToken);
+        Assert.NotNull(result);
+        Assert.Equal(2, result.Contents.Count);
+        Assert.Equal("42", ((TextResourceContents)result.Contents[0]).Text);
+        Assert.Equal("43", ((TextResourceContents)result.Contents[1]).Text);
+    }
+
+    [Fact]
+    public async Task CanReturnDataContent()
+    {
+        Mock<IMcpServer> mockServer = new();
+        McpServerResourceTemplate resource = McpServerResourceTemplate.Create((IMcpServer server) =>
+        {
+            Assert.Same(mockServer.Object, server);
+            return new DataContent(new byte[] { 0, 1, 2 }, "application/octet-stream");
+        }, new() { Name = "Test" });
+        var result = await resource.ReadAsync(
+            new RequestContext<ReadResourceRequestParams>(mockServer.Object) { Params = new() { Uri = "resource://Test" } },
+            TestContext.Current.CancellationToken);
+        Assert.NotNull(result);
+        Assert.Single(result.Contents);
+        Assert.Equal(Convert.ToBase64String(new byte[] { 0, 1, 2 }), ((BlobResourceContents)result.Contents[0]).Blob);
+        Assert.Equal("application/octet-stream", ((BlobResourceContents)result.Contents[0]).MimeType);
+    }
+
+    [Fact]
+    public async Task CanReturnCollectionOfAIContent()
+    {
+        Mock<IMcpServer> mockServer = new();
+        McpServerResourceTemplate resource = McpServerResourceTemplate.Create((IMcpServer server) =>
+        {
+            Assert.Same(mockServer.Object, server);
+            return new List<AIContent>()
+            {
+                new TextContent("hello!"),
+                new DataContent(new byte[] { 4, 5, 6 }, "application/json"),
+            };
+        }, new() { Name = "Test" });
+        var result = await resource.ReadAsync(
+            new RequestContext<ReadResourceRequestParams>(mockServer.Object) { Params = new() { Uri = "resource://Test" } },
+            TestContext.Current.CancellationToken);
+        Assert.NotNull(result);
+        Assert.Equal(2, result.Contents.Count);
+        Assert.Equal("hello!", ((TextResourceContents)result.Contents[0]).Text);
+        Assert.Equal(Convert.ToBase64String(new byte[] { 4, 5, 6 }), ((BlobResourceContents)result.Contents[1]).Blob);
+        Assert.Equal("application/json", ((BlobResourceContents)result.Contents[1]).MimeType);
+    }
+
+    private sealed class MyService;
+
+    private class DisposableResourceType : IDisposable
+    {
+        public static int Disposals { get; private set; }
+
+        public void Dispose() => Disposals++;
+
+        [McpServerResourceTemplate(UriTemplate = "test://static/resource/instanceMethod")]
```

Comment:
If we're going to delineate between [McpServerResource] and [McpServerResourceTemplate], should we throw a startup error if there's no templating in the UriTemplate? I know templated resources show up in a different listing than non-templated resources, so it's slightly different behavior to go non-templated. But you'd still reading these resources the same way, so it's not *that* different, and more likely what you want.

If you really want to have a non-templated resource in the templated listing, I think you should have to drop down and call the lower-level WithListResourceTemplatesHandler API that we already have.